odilo.py

```diff
@@ -147,18 +147,37 @@ def refresh_creds(self, credential):
         response = self.token_post(
             self.TOKEN_ENDPOINT,
             dict(grant_type="client_credentials"),
-            allowed_response_codes=[200]
+            allowed_response_codes=[200, 400]
         )
-        data = response.json()
+
+        # If you put in the wrong URL, this is where you'll run into
+        # problems, so it's useful to give a helpful error message if
+        # Odilo doesn't provide anything more specific.
+        generic_error = "%s may not be the right base URL. Response document was: %r" % (
+            self.library_api_base_url, response.content
+        )
+        generic_exception = BadResponseException(
+            self.TOKEN_ENDPOINT, generic_error
+        )
+
+        try:
+            data = response.json()
+        except ValueError:
+            raise generic_exception
 
         if response.status_code == 200:
             self._update_credential(credential, data)
             self.token = credential.credential
+            return
         elif response.status_code == 400:
             if data and 'errors' in data and len(data['errors']) > 0:
                 error = data['errors'][0]
-                message = ('description' in error and error['description']) or ''
-                raise BadResponseException(message)
+                if 'description' in error:
+                    message = error['description']
+                else:
+                    message = generic_error
+                raise BadResponseException(self.TOKEN_ENDPOINT, message)
```

Comment:
This is just to make sure the generic error is used instead of the empty string if Odilo (or whoever) sends back a 400 response code and JSON but no specific error message.